### PR TITLE
[DebugInfo] Handle trailing empty blocks when seeking prologue_end spot

### PR DESCRIPTION
**Pull request by jmorse**
The optimiser will produce empty blocks that are unconditionally executed according to the CFG -- while it may not be meaningful code, and won't get a prologue_end position, we need to not crash on this input.

The fault comes from assuming that there's always a next block with some instructions in it, that will eventually produce some meaningful control flow to stop at -- in the given reproducer in issue #117206 this isn't true, because the function terminates with `unreachable`. Thus, I've refactored the "get next instruction logic" into a helper that'll step through all blocks and terminate if there aren't any more.

Reproducer from @aeubanks

### CODE REVIEW

File: llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp [hunk 1]
```diff
@@ -2216,25 +2244,8 @@ findPrologueEndLoc(const MachineFunction *MF) {
...
-    // If we've already fallen through into a loop, don't fall through
-    // further, use a backup-location.
-    if (CurBlock->pred_size() > 1)
-      break;
-
-    // Fall-through from entry to the next block. This is common at -O0 when
-    // there's no initialisation in the function. Bail if we're also at the
-    // end of the function.
-    if (++CurBlock == MF->end())
+    if (!getNextInst())
```
Comment (OCHyams) [inline llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp:2247]: I get what this is doing but from the name `getNextInst()` I think it sounds more like it's doing `auto NextInst = std::next(CurInst)` as above. Maybe `getNextBlockInst` ? That doesn't sound great either. `findNextBlock`? Ymmv, feel free to ignore.

File: llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp [hunk 2]
```diff
@@ -2192,14 +2192,42 @@ findPrologueEndLoc(const MachineFunction *MF) {
...
   auto CurInst = CurBlock->begin();
-  while (true) {
+
+  // Find the initial instruction, we're guaranteed one by the caller.
+  while (CurBlock->empty())
+    CurInst = (++CurBlock)->begin();
+
+  // Helper function for stepping through the initial sequence of
+  // unconditionally executed instructions. The optimiser will also chuck in
+  // sequences of empty blocks alas.
```
Comment (OCHyams) [inline llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp:2202]: I wonder if "The optimiser will also chuck in sequences of empty blocks alas." is slightly too informal (could cause trouble for folks with English as a 2nd language?). Also probably worth moving this part of the into the lambda body as it seems more like an implementation detail?

File: llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp [hunk 3]
```diff
@@ -2192,14 +2192,42 @@ findPrologueEndLoc(const MachineFunction *MF) {
   // better off synthesising an early prologue_end.
   auto CurBlock = MF->begin();
   auto CurInst = CurBlock->begin();
-  while (true) {
+
+  // Find the initial instruction, we're guaranteed one by the caller.
+  while (CurBlock->empty())
+    CurInst = (++CurBlock)->begin();
+
```
Comment (OCHyams) [inline llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp:2199]: `// Find the initial instruction, we're guaranteed one by the caller.` Can we assert `CurInst != CurBlock->end()` here to check that?

**Review by OCHyams — APPROVED** [review]
LGTM with some nits

File: llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp (hunk 2 above)
Comment (jmorse) [reply llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp:2202]: Ah yeah, this is written in my default "Jeremy-flippancy" dialect

**Review by OCHyams — APPROVED** [review]
Still LGTM

**Pull request merged**